Automatically generate changelog revision text if the changelog was not hand-crafted - do not commit if unchanged

weierophinney opened the pull request:
|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| BC Break      | code, yes; UI and usage, no
| New Feature   | yes and no
| RFC           | yes

### Description

When rolling out automatic-releases to various repositories, we noticed an issue with the workflow for documentation, mainly that our docs-build action only builds on pushes to the "master" branch, we are trying to eliminate. This has meant manually pushing the current release branch to the "master" branch on the repository whenever there are changes.

@Ocramius suggested that we merge documentation changes just like bugfixes: assign them to milestones, and build on release. The objections raised were:

- We would have unchanged packages from one version to the next, as docs are excluded via `.gitattributes`. We decided this was not really an issue; most people are not updating regularly, and a package that introduces zero code changes is generally welcome as it represents no risk.

- We would have empty changelogs and release announcements.

This latter was something I realized we could address in the automatic-releases tooling, as:

- We already generate additional release notes via jwage/changelog-generator.
- The phly/keep-a-changelog tooling does not require that entries are in a specific format unless you want to use its `entry:*` commands; the only requirements are the version headings.

As such, this patch refactors how changelogs are generated for release notes, and then committed back to the repository.

In particular:

- It introduces `Laminas\AutomaticReleases\Changelog\ChangelogReleaseNotes`, which is an entity that composes the changelog contents, and optionally a phly/keep-a-changelog `ChangelogEntry` instance. It has behavior:
  - `requiresUpdatingChangelogFile()` returns true if a `ChangelogEntry` is composed AND its contents differ from those stored in the `ChangelogReleaseNotes` instance.
  - `writeChangelogFile(string $changelogFile)` will write the release notes contents to the specified changelog file, per the metadata in the `ChangelogEntry` instance composed.
  - `merge(ChangelogReleaseNotes $next): ChangelogReleaseNotes` will clone itself and merge `$next` into it, returning the clone.

- I updated `CreateReleaseText` to have it return a `ChangelogReleaseNotes` instance.

- I updated `CreateReleaseTextViaKeepAChangelog` to now mimics how the phly/keep-a-changelog library parses a changelog file when parsing string contents representing a full changelog as pulled from the originating branch, in order to create a `ChangelogEntry` that details where the original is found in the file, and how much space it takes. Additionally, it then sets the date for the version, and checks to see if the contents have changed from the template; if not, it clears them before passing them to the `ChangelogReleaseNotes` instance it creates.

- I updated `CreateReleaseThroughChangelog` to perform some minor markdown formatting changes to better accommodate being injected in a CHANGELOG.md file; it also now returns a `ChangelogReleaseNotes` instance (without a `ChangelogEntry` composed).

- I updated `CommitReleaseChangelog` to accept a `ChangelogReleaseNotes` as the first argument; implementations now use it to actually write the `CHANGELOG.md` file prior to committing and pushing it.

- I renamed `ConcatenateMultipleReleaseTexts` to `MergeMultipleReleaseNotes` to better describe what it now does (it uses the `merge()` operation of returned `ChangelogReleaseNotes` instances, returning the final product).

- I updated the logic of `ReleaseCommand` to generate release text first, then commit it.

## tl;dr

When the "release" command of automatic-releases now runs, it will:

- Always generate release notes using jwage/changelog-generator.
- **IF** a `CHANGELOG.md` file is present, it then:
  - Checks to see if it contains any changes for the target version. If not, it replaces the content with the generated release notes.
  - Otherwise, it appends the content with the generated release notes.

The upshot is that documentation maintainers can:

- Add documentation patches to the appropriate milestone.
- Merge those changes.
- Close the milestone.

and get a release issued, which will then trigger builds for documentation.

## Final note

We will need to use the `ORGANIZATION_ADMIN_TOKEN` for the value of the `GITHUB_TOKEN` on the `Release` step of the workflow in order to trigger the release workflow event, and thus the documentation build.

weierophinney committed:
fix: revise format of generated changelog

- Use a third-level heading for the opening statement, and chnage from
  "Release ..." to "Release Notes for ..."

- Rewrite '----' and '====' style headings to appropriate `#` style headings (consistency)

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: merge changelog.md contents with generated changelog

This patch accomplishes several goals:

- When a changelog entry from a CHANGELOG.md file does not contain any
  changes, its contents are removed.
- The contents from jwage changelog generator either replace those in
  for the entry in CHANGELOG.md (if it had no changes), or append to it,
  ALWAYS. These are then COMMITTED to the file.

To accomplish this, I needed to do the following:

- I created the class `ChangelogReleaseNotes`. It aggregates changelog
  contents, and, optionally, a phly/keep-a-changelog `ChangelogEntry`
  instance. It defines three behavior methods in addition to its
  `contents()` value method:

  - `requiresUpdatingChangelogFile()`, which returns true if there is a
    changelog entry instance AND its contents differ from those of the
    `ChangelogReleaseNotes` instance itself.
  - `writeChangelogFile(string $changelogFile)` will write the contents
    aggregated in the instance to the specified changelog file, per the
    composed `$changelogEntry` instance
  - `merge(ChangelogReleaseNotes $next): ChangelogReleaseNotes` will
    create a new instance that merges the current with the one provided.
    The contents of the one provided are appended to the original using
    a separator, and if either compose a `ChangelogEntry` instance, it
    will be used. (If both define one, this is considered undefined
    behavior.)

- I updated the CreateReleaseText interface to return a
  `ChangelogReleaseNotes`  instance.

- I updated the `CreateReleaseTextViaKeepAChangelog` to compose a
  `Clock` instance, and to inline logic for creating a `ChangelogEntry`
  instance when parsing the `CHANGELOG.md` file for the related version.
  Once it has, it removes the entry if it is unchanged since
  initialization, and sets the date for the release before seeding a
  `ChangelogReleaseNotes` instance to return.

- I updated `CreateReleaseTextThroughChangelog` to create and return a
  `ChangelogReleaseNotes` instance with the text it creates.

- I renamed `ConcatenateMultipleReleaseTexts` to
  `MergeMultipleReleaseNotes`, and have it now using `array_reduce` on its
  aggregated items to `merge()` each result with the previous, returning
  the merged `ChangelogReleaseNotes` instance.

- I updated the `CommitReleaseChangelog` interface to accept a
  `ChangelogReleaseNotes` instance as the first argument. It then test to
  see if it has changes via `requiresUpdatingChangelogFile()`, and, if so,
  it uses its `writeChangelogFile()` method to write the changes before
  committing them.

- In `ReleaseCommand`, I inverted the release text creation and commit
  operations, and capture the `ChangelogReleaseNotes` for use with both
  the commit and the release operations.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: updates CHANGELOG.md

Adds information detailing user-facing changes of #58.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on src/Changelog/ChangelogReleaseNotes.php:
```diff
@@ -0,0 +1,86 @@
+use RuntimeException;
+use Webmozart\Assert\Assert;
+
+class ChangelogReleaseNotes
```
I'd love for this to be `@psalm-immutable`, but we do perform filesystem interactions below.

It is still possible to achieve that by making the method that performs side effects `static`, but that isn't the end of the world, so I'll leave it up to the author's discretion

weierophinney replied:
So, I tried doing this just now, and the problem I ran into is that `ChangelogEntry` is mutable, and Psalm doesn't like having an externally mutable instance inside an immutable object. I tried cloning it, but it somehow fails to see that there's a clone created internally, and that the instance is not exposed publicly at any point (there's no getter/setter for it).

I think I'll move forward with it, but will need to suppress the `ImpurePropertyAssignment` warnings, knowing that they were being raised incorrectly. (Just letting you know in advance.)

Ocramius replied:
Ah, I see, `ChangelogEntry` is indeed external. Missed that, sorry!

Overall, I'm OK with suppressing those for now, so this version of the code should be fine.

Ocramius commented on src/Application/Command/ReleaseCommand.php:
```diff
@@ -84,22 +84,28 @@ public function execute(InputInterface $input, OutputInterface $output): int
             $repositoryPath
         );
 
+        ($this->commitChangelog)($changelogReleaseNotes, $repositoryPath, $releaseVersion, $releaseBranch);
```
I like how this now receives a pre-computed changelog 👍

weierophinney replied:
I thought you might. :smile:

weierophinney replied:
This made me happy as well. The "commit, then create" from before worked, and was necessary based on what we were doing, but this makes sooooo much more sense.

Ocramius commented on src/Changelog/CommitReleaseChangelogViaKeepAChangelog.php:
```diff
@@ -90,104 +86,4 @@ public function __invoke(
         ($this->push)($repositoryDirectory, $sourceBranch->name());
     }
-
-    private function updateChangelog(string $changelogFile, string $versionString): bool
```
Excellent refactoring 💪

Ocramius commented on src/Github/CreateReleaseTextThroughChangelog.php:
```diff
@@ -73,4 +80,36 @@ private function markdownLink(string $text, UriInterface $uri): string
+        $lines         = explode("\n", trim($changelog));
+        $lineCount     = count($lines);
+        $linesToRemove = [];
+        for ($i = 0; $i < $lineCount; $i += 1) {
```
This code block is really hard to understand. I suppose most of it is tested well, according to mutation testing results, but it is really making the reader's head spin.

Refactoring can be deferred, since the code is functionally pure, but we need comments about what is being done here.

samsonasik replied:
how about using `foreach` instead ? 

```suggestion
        foreach ($lines as $i => $line) {
```

We can eliminate usage of `$lineCount     = count($lines);`

Ocramius commented on src/Github/CreateReleaseTextThroughChangelog.php:
```diff
@@ -73,4 +80,36 @@ private function markdownLink(string $text, UriInterface $uri): string
+            }
+
+            $previousLine = $lines[$i - 1];
+            if (empty($previousLine)) {
```
What happens with `$i = 0`? I expect a notice in the line above 🤔

Ocramius replied:
Fixed by conditional introduced above :+1:

Ocramius commented on src/Github/CreateReleaseTextViaKeepAChangelog.php:
```diff
@@ -84,4 +128,78 @@ private function fetchChangelogContentsFromBranch(
+        $boundaryRegex = '/^(?:## (?:\d+\.\d+\.\d+|\[\d+\.\d+\.\d+\])|\[.*?\]:\s*\S+)/i';
+        $regex         = sprintf('/^## (?:%1$s|\[%1$s\])/i', preg_quote($version));
+
+        foreach (explode("\n", $contents) as $index => $line) {
```
Similar to other changelog normalisation logic, this is really hard to follow.

Can defer refactoring, but need comments about what is being done in this method.

Ocramius commented on src/Github/MergeMultipleReleaseNotes.php:
```diff
@@ -33,19 +32,24 @@ public function __invoke(
+        );
+
+        Assert::isInstanceOf($releaseNotes, ChangelogReleaseNotes::class);
+        Assert::notEmpty($releaseNotes->contents());
```
Is this assertion needed? I think `->contents()` already returns a `non-empty-string`

Ocramius commented on test/unit/Changelog/ChangelogReleaseNotesTest.php:
```diff
@@ -0,0 +1,222 @@
+use function sprintf;
+use function sys_get_temp_dir;
+
+class ChangelogReleaseNotesTest extends TestCase
```
Consider adding `@covers`: makes mutation testing more precise on the SUT

Ocramius commented on test/unit/Github/CreateChangelogTextTest.php:
```diff
@@ -10,12 +10,14 @@
 {
     public function testGeneratedReleaseText(): void
     {
+        /** @psalm-var GenerateChangelog&MockObject $generateChangelog */
```
This inline type shouldn't be needed: phpunit mocking utilities already have the correct annotations

weierophinney replied:
This was in part for intelephense code completion.

Ocramius replied:
Doesn't intelephense read something like `.phpstorm.meta.php`?

Ocramius reviewed (changes requested):
One potential bug, some space for refactoring, otherwise requires commenting on tricky logic that parses release notes.

I think this needs only quick adjustments, and then we can release it as `1.3.0`

samsonasik commented on src/Github/CreateReleaseTextThroughChangelog.php:
```diff
@@ -73,4 +80,36 @@ private function markdownLink(string $text, UriInterface $uri): string
+    private function normalizeChangelogHeadings(string $changelog): string
+    {
+        $lines         = explode("\n", trim($changelog));
+        $lineCount     = count($lines);
```
can be removed if next `for` changed to `foreach`

michalbundyra commented on src/Github/CreateReleaseText.php:
```diff
@@ -13,15 +14,14 @@ interface CreateReleaseText
         BranchName $sourceBranch,
         string $repositoryDirectory
-    ): string;
+    ): ChangelogReleaseNotes;
```
BC Break 😱 Technically then this feature should go to v2.0

Ocramius replied:
This package has `"type": "application"` and is not published to packagist either - we can break internals as much as we want (and feel free to do so!)

weierophinney committed:
qa: mark ChangelogReleaseNotes as immutable

Requires making the `writeChangelogFile()` method static.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: document logic of parsing and normalization of release notes

Details the algorithms for each, and what they are expected to
accomplish.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove unnecessary assertion

Already guaranteed by `ChangelogReleaseNotes` method definition.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: testing improvements

- Adds `@covers` annotation to `ChangelogReleaseNotesTest`
- Removes unnecessary variable annotation in `CreateChangelogTextTest`

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius reviewed (approved):
LGTM :ship:

Ocramius commented:
Ah, can't merge this due to mutation score going down quite a lot. This is most likely due to the complex logic in the changelog text generation/extraction code: we can either test it more carefully (more string examples, effectively), or rewrite bits of it to be more functional-oriented, and removing prodcedural/iteration bits that are more mutation-prone (yes, FP has the advantage of reducing code paths :D )

weierophinney committed:
qa: fix @covers annotation

Requires fully qualified class name, not one derived from import
statements.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: increase test coverage

Updates test cases to ensure all paths of parsers and normalizers are
tested.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Mutation score fixed, @Ocramius :smile:

weierophinney committed:
docs: provide additional workflow example enabling release events

To simplify setup in our own repos, I'm adding another example workflow
file that uses the organization admin token in the release step, to
allow triggering a release workflow event (which will be necessary for
us to trigger documenation builds).

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on examples/.github/workflows/release-on-milestone-closed-triggering-release-event.yml:
```diff
@@ -0,0 +1,57 @@
+# https://help.github.com/en/categories/automating-your-workflow-with-github-actions
```
This example needs a better comment

Ocramius reviewed (approved):
LGTM - the documentation of the new example can wait a bit
